google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionClientTest.java

```diff
@@ -193,6 +195,73 @@ public void onSessionCreateFailure(Throwable t, int createFailureForSessionCount
+   * calls to this method will still distribute the total set of sessions over all available
+   * channels.
+   */
+  @SuppressWarnings("unchecked")
+  @Test
+  public void batchCreateSessionsDistributesMultipleRequestsOverChannels() {
+    DatabaseId db = DatabaseId.of(dbName);
+    final String sessionName = dbName + "/sessions/s%d";
+    final Map<String, String> labels = Collections.<String, String>emptyMap();
+    when(spannerOptions.getSessionLabels()).thenReturn(labels);
+    final Set<Long> usedChannelHintss = Collections.synchronizedSet(new HashSet<Long>());
+    when(rpc.batchCreateSessions(
+            Mockito.eq(dbName), Mockito.anyInt(), Mockito.eq(labels), Mockito.anyMap()))
+        .then(
+            new Answer<List<com.google.spanner.v1.Session>>() {
+              @Override
+              public List<com.google.spanner.v1.Session> answer(InvocationOnMock invocation)
+                  throws Throwable {
+                Map<SpannerRpc.Option, Object> options = invocation.getArgumentAt(3, Map.class);
+                Long channelHint = (Long) options.get(SpannerRpc.Option.CHANNEL_HINT);
+                usedChannelHintss.add(channelHint);
+                int sessionCount = invocation.getArgumentAt(1, Integer.class);
+                List<com.google.spanner.v1.Session> res = new ArrayList<>();
+                for (int i = 1; i <= sessionCount; i++) {
+                  res.add(
+                      com.google.spanner.v1.Session.newBuilder()
+                          .setName(String.format(sessionName, i))
+                          .putAllLabels(labels)
+                          .build());
+                }
+                return res;
+              }
+            });
+
+    final AtomicInteger returnedSessionCount = new AtomicInteger();
+    SessionConsumer consumer =
+        new SessionConsumer() {
+          @Override
+          public void onSessionReady(SessionImpl session) {
+            assertThat(session.getName()).startsWith(dbName + "/sessions/s");
+            returnedSessionCount.incrementAndGet();
+            session.close();
+          }
+
+          @Override
+          public void onSessionCreateFailure(Throwable t, int createFailureForSessionCount) {}
+        };
+    final int numSessions = 10;
+    final int numBatches = spannerOptions.getNumChannels() * 2;
+    try (SessionClient client = new SessionClient(spanner, db, new TestExecutorFactory())) {
+      for (int batch = 0; batch < numBatches; batch++) {
+        client.asyncBatchCreateSessions(numSessions, false, consumer);
+      }
+    }
+    assertThat(returnedSessionCount.get()).isEqualTo(numSessions * numBatches);
+    assertThat(usedChannelHintss.size()).isEqualTo(spannerOptions.getNumChannels() * 2);
+    List<Long> expectedChannels = new ArrayList<>();
+    for (long l = 0; l < spannerOptions.getNumChannels() * 2; l++) {
+      expectedChannels.add(l);
+    }
+    assertThat(usedChannelHintss).containsExactlyElementsIn(expectedChannels);
```

Comment:
Assume that `spannerOptions.getNumChannels()` returns 4. So this means that we only have 4 available channels. 

But why `expectedChannels` can be `[0,1,2,3,...,7]`?

Reply:
I see. 

https://github.com/googleapis/java-spanner/blob/9e5a1cdaacf71147b67681861f063c3276705f44/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionClient.java#L175

There is an increasing counter in SessionClient. Every time when we call `createSession`, `BatchCreateSessionsRunnable`, and `sessionWithId`, `sessionChannelCounter` increases to make sure they use different channels.

Reply:
Java gRPC uses a channel hint to indicate that a call should be linked to a certain channel. That channel hint can be any `long` value, and gRPC will automatically bring that back to a certain channel number. The current implementation is `channelHint % numChannels`, but the `SessionClient` should not really depend on that (or care about that). The API specifies that any `long` value is OK, so the `SessionClient` just assigns a continuously increasing `long`.